### DISCUSSION TITLE
Auto-scale line values with negative and positive

### DISCUSSION THREAD

**Discussion by eric-nextsense** [patrykandpatrick/vico#816] [Questions] [answered]
### Question

I am trying to find the best way to auto-scale my graph that can contain both positive and negative values. I have a known range that is possible (-91000 to 91000), but i want to show only the current range in the graph. This changes while graphing.

I tried to use adaptiveY values, but it does not seem to work well when there are values under 0. I am not sure how this code works exactly, but maybe this ".coerceAtLeast(0.0)" tries to center it on 0, which i do not want? I want completely floating based on the current values.

auto mode is based on 0 so it did not work well either.

    public fun adaptiveYValues(yFraction: Float, round: Boolean = false): AxisValueOverrider =
      object : AxisValueOverrider {
        private val Double.conditionallyRoundedToNearest
          get() = if (round) roundedToNearest else this

        init {
          require(yFraction > 0f)
        }

        override fun getMinY(minY: Double, maxY: Double, extraStore: ExtraStore): Double {
          val difference = abs(getMaxY(minY, maxY, extraStore) - maxY)
          return (minY - difference).conditionallyRoundedToNearest.coerceAtLeast(0.0)
        }

        override fun getMaxY(minY: Double, maxY: Double, extraStore: ExtraStore): Double =
          if (minY == 0.0 && maxY == 0.0) 1.0 else (yFraction * maxY).conditionallyRoundedToNearest
      }
  }

I did some experimentation and this code seems to work for my needs at least with yFraction = 1.0f, does it makes sense? Maybe it could be added in the library?

fun fullyAdaptiveYValues(yFraction: Float, round: Boolean = false): AxisValueOverrider =
    object : AxisValueOverrider {
        private val Double.conditionallyRoundedToNearest
            get() = if (round) roundedToNearest else this

        init {
            require(yFraction > 0f)
        }

        override fun getMinY(minY: Double, maxY: Double, extraStore: ExtraStore): Double {
            val difference = abs(getMaxY(minY, maxY, extraStore) - maxY)
            return (minY - difference).conditionallyRoundedToNearest
        }

        override fun getMaxY(minY: Double, maxY: Double, extraStore: ExtraStore): Double =
            if (minY.roundedToNearest == maxY.roundedToNearest) maxY + 1 else (yFraction * maxY).conditionallyRoundedToNearest
    }

**Comment by patrickmichalik** [comment, marked as answer]
Hello! `getMinY` and `getMaxY` are passed the data _y_ range. To obtain the desired behavior, you can create an `AxisValueOverrider` that leaves the values that it receives either mostly or entirely unmodified.

If the data’s minimum and maximum _y_ values always differ, this should suffice:

```kt
object : AxisValueOverrider {
  override fun getMinY(minY: Double, maxY: Double, extraStore: ExtraStore) = minY

  override fun getMaxY(minY: Double, maxY: Double, extraStore: ExtraStore) = maxY
}
```

If the data’s minimum and maximum _y_ values may be equal, you should ensure that the final _y_ range is not zero-length. Here’s a possible approach:

```kt
object : AxisValueOverrider {
  override fun getMinY(minY: Double, maxY: Double, extraStore: ExtraStore) =
    if (minY == maxY) minY - 1 else minY

  override fun getMaxY(minY: Double, maxY: Double, extraStore: ExtraStore) =
    if (minY == maxY) maxY + 1 else maxY
}
```

We’ll look into adding such a built-in implementation.

This `AxisValueOverrider` generally produces the same result as your custom one with `yFraction` set to 1 and `round` set to `false`, but in a more direct fashion. I say “generally” because in your `AxisValueOverrider`, `minY` and `maxY` are rounded in the equality check. I presume this is an error—it could cause somewhat inconsistent behavior, and your description doesn’t point to this behavior being desired.

Regarding the use of `coerceAtLeast(0.0)` in the `AxisValueOverrider` returned by `adaptiveYValues`, this appears to be a mistake on our part. The intended behavior of this `AxisValueOverrider` is described in the `adaptiveYValues` documentation. We’ll correct the logic accordingly.

**Reply by eric-nextsense** [reply]
Thank you this is helpful!

I had tried with a straight equality and it did not work. I think the float comparison failed as they were probably very near but not exactly the same. The graph would not display anything when that happened. I suppose the min and max were rounded to generate the axis and they ended up the same.

Rounding them to compare worked fine and is good enough for my scenario.

**Reply by patrickmichalik** [reply]
You’re welcome, @eric-nextsense!

Got it. If there are instances in which `minY` and `maxY` are very close to each other but aren’t equal, then `==` won’t suffice. The final _y_ range isn’t rounded, but an exceedingly narrow range may cause floating-point precision issues to crop up during drawing, leading to unexpected behavior. Comparing `maxY - minY` to a predefined threshold may, however, be preferable to using `roundToNearest`. There are two reasons.

First, the `roundToNearest` solution produces such inconsistencies:

- For `minY` = 1.1 and `maxY` = 1.2, the condition is true, and the final _y_ range is [0.1, 2.2].
- For `minY` = 1.4 and `maxY` = 1.5, the condition is false, and the final _y_ range is [1.4, 1.5].

Despite the data _y_ range having the same length, 0.1, in both cases, there’s a significant difference in behavior.

Second, with `roundToNearest`, there are still instances in which nearly identical values are treated as different, and a nearly zero-length final _y_ range is produced. This occurs, for example, with `minY` = 1.499999999999999 and `maxY` = 1.5, which are rounded to different whole numbers.

A threshold comparison would resolve both of these issues without relying on equality.

**Comment by eric-nextsense** [comment]
Implemented it that way and it works perfectly now. I indeed had some issues with the rounding in some cases as you pointed out.

```
fun Double.nearValue(other: Double) : Boolean {
    return (abs(this - other) < 0.01)
}

fun fullyAdaptiveYValues(): AxisValueOverrider =
    object : AxisValueOverrider {
        override fun getMinY(minY: Double, maxY: Double, extraStore: ExtraStore) =
            if (minY.nearValue(maxY)) minY - 1 else minY

        override fun getMaxY(minY: Double, maxY: Double, extraStore: ExtraStore) =
            if (minY.nearValue(maxY)) maxY + 1 else maxY
    }
```

**Reply by patrickmichalik** [reply]
Looks good! I’m glad to hear that this works as desired.
